lib/src/Base/Algo/SimplicialCubature.cxx

```diff
@@ -0,0 +1,852 @@
+    {
+      volume[K] = VI;
+      computeRuleValueAndError(dimension, flatVertices[K], VI, outputDimension, F, G, W, evalBudget, value, error);
+      VLS[K] = value;
+      AES[K] = error;
+      VL += VLS[K];
+      AE += AES[K];
+      NV += nodeNumber;
+    }
+
+    NV += DFCOST;
+    flatSize += NEW - 1;
+
+    // Check for error termination.
+    FL = 0;
+    for (UnsignedInteger i = 0; i < outputDimension; ++ i)
+      if (AE[i] > std::max(EA, ER*abs(VL[i])))
+      {
+        FL = 1;
+        break;
+      }
+  }
+  return VL;
+}
+
+Point SimplicialCubature::computePermutationSums(const UnsignedInteger dimension, const Sample & simplexVertices,
+                                                 const UnsignedInteger outputDimension, const Function & function,
+                                                 const Point & GConst) const
+{
+  Point result(outputDimension, 0.0);
+  Point G(GConst);
+  std::sort(G.begin(), G.end(), std::greater<Scalar>());
+
+  Bool pr = true;
+  // Compute integrand value for permutations of G
+  while (pr)
+  { 
+    // TODO: better matrix conversion ?
+    Matrix VERTEXM(simplexVertices.getDimension(), simplexVertices.getSize());
+    for (UnsignedInteger i = 0; i < simplexVertices.getSize(); ++ i)
+      for (UnsignedInteger j = 0; j < simplexVertices.getDimension(); ++ j)
+        VERTEXM(j, i) = simplexVertices(i, j);
+
+    result += function(VERTEXM * G);
+    pr = false;
+
+    for (UnsignedInteger I = 1; I < dimension+1; ++ I)
+    {
+      const Scalar GI = G[I];
+      if (G[I-1] > GI)
+      {
+        UnsignedInteger IX = I - 1;
+        UnsignedInteger LX = 0;
+        for (UnsignedInteger L = 0; L < ((IX+2)/2); ++ L)
+        {
+          const Scalar GL = G[L];
+          if (GL <= GI)
+            --IX;
+          G[L] = G[I-L-1];
+          G[I-L-1] = GL;
+          if (G[L] > GI)
+            LX = L;
+        }
+        if (G[IX] <= GI)
+          IX = LX;
+        G[I] = G[IX];
+        G[IX] = GI;
+        pr = true;
+        break;
+      }
+    }
+  }
+  return result;
+}
+
+void SimplicialCubature::computeRuleValueAndError(const UnsignedInteger dimension, const Sample & simplexVertices,
+                                                  const Scalar volume, const UnsignedInteger outputDimension,
+                                                  const Function & function, const Matrix & G, const Matrix & W,
+                                                  const Indices & evalBudget, Point & value, Point & error) const
+{
+  Scalar RTMN = 1e-1;
+  Scalar SMALL = SpecFunc::Precision;
+  Scalar ERRCOF = 8;
+
+  const UnsignedInteger WTS = W.getNbRows();
+  const UnsignedInteger RLS = W.getNbColumns();
+  Matrix RULE(outputDimension, RLS);
+
+  for (UnsignedInteger K = 0; K < WTS; ++ K)
+    if (evalBudget[K] > 0)
+    {
+      Point sym(computePermutationSums(dimension, simplexVertices, outputDimension, function, G.getColumn(K)*Point(1, 1.0)));
+      // TODO: better solution?
```

Comment:
You should probably unroll the computation to improve the performance. Something like:
```
for (UnsignedInteger i = 0; i < outputDimension; ++ i)
{
  const Scalar scaledSymI = volume * sym[i];
  for (UnsignedInteger j = 0; j < RLS; ++ j)
    RULE(i, j) += scaledSymI * W(K, j);
}